libp2p/protocols/relayv2/relayv2.nim

```diff
@@ -0,0 +1,860 @@
+  if msg.peer.isSome():
+    var ppb = initProtoBuffer()
+    ppb.write(1, msg.peer.get().peerId)
+    for ma in msg.peer.get().addrs:
+      ppb.write(2, ma.data.buffer)
+    ppb.finish()
+    pb.write(2, ppb.buffer)
+  if msg.limit.isSome():
+    var lpb = initProtoBuffer()
+    let limit = msg.limit.get()
+    lpb.write(1, limit.duration)
+    lpb.write(2, limit.data)
+    lpb.finish()
+    pb.write(3, lpb.buffer)
+  if msg.status.isSome():
+    pb.write(4, msg.status.get().ord.uint)
+
+  pb.finish()
+  pb
+
+proc decodeStopMessage(buf: seq[byte]): Option[StopMessage] =
+  var
+    msg: StopMessage
+    msgTypeOrd: uint32
+    pbPeer: ProtoBuffer
+    pbLimit: ProtoBuffer
+    statusOrd: uint32
+    peer: RelayV2Peer
+    limit: RelayV2Limit
+    rVoucher: ProtoResult[bool]
+    res: bool
+
+  let
+    pb = initProtoBuffer(buf)
+    r1 = pb.getRequiredField(1, msgTypeOrd)
+    r2 = pb.getField(2, pbPeer)
+    r3 = pb.getField(3, pbLimit)
+    r4 = pb.getField(4, statusOrd)
+
+  if r1.isErr() or r2.isErr() or r3.isErr() or r4.isErr():
+    return none(StopMessage)
+
+  if r2.get() and
+     (pbPeer.getRequiredField(1, peer.peerId).isErr() or
+      pbPeer.getRepeatedField(2, peer.addrs).isErr()):
+    return none(StopMessage)
+
+  if r3.get() and
+     (pbLimit.getField(1, limit.duration).isErr() or
+      pbLimit.getField(2, limit.data).isErr()):
+    return none(StopMessage)
+
+  msg.msgType = StopMessageType(msgTypeOrd)
+  if r2.get(): msg.peer = some(peer)
+  if r3.get(): msg.limit = some(limit)
+  if r4.get(): msg.status = some(RelayV2Status(statusOrd))
+  some(msg)
+
+# Protocols
+
+proc createHopMessage(
+    rv2: RelayV2,
+    pid: PeerID,
+    expire: DateTime): Result[HopMessage, CryptoError] =
+  var msg: HopMessage
+  let
+    expireUnix = expire.toTime.toUnix.uint64
+    v = Voucher(relayPeerId: rv2.switch.peerInfo.peerId,
+                  reservingPeerId: pid,
+                  expiration: expireUnix)
+    sv = ? SignedVoucher.init(rv2.switch.peerInfo.privateKey, v)
+    ma = MultiAddress.init("/p2p/" & $rv2.switch.peerInfo.peerId).tryGet()
+  msg.reservation = some(Reservation(expire: expireUnix,
+                         addrs: rv2.switch.peerInfo.addrs.mapIt(it & ma),
+                         svoucher: some(? sv.encode)))
+  msg.limit = some(rv2.limit)
+  msg.msgType = HopMessageType.Status
+  msg.status = some(Ok)
+  return ok(msg)
+
+proc handleReserve(rv2: RelayV2, conn: Connection) {.async, gcsafe.} =
+  let pid = conn.peerId
+
+  if conn.isCircuitRelay():
+    trace "reservation attempt over relay connection", pid
+    await handleHopError(conn, RelayV2Status.PermissionDenied)
+    return
+
+  let expire = now().utc + rv2.reservationTTL
+  rv2.rsvp[pid] = expire
+
+  trace "reserving relay slot for", pid
+
+  let msg = rv2.createHopMessage(pid, expire)
+  if msg.isErr():
+    trace "error signing the voucher", error = error(msg), pid
+    # conn.reset()
+    return
+  let pb = encodeHopMessage(msg.get())
+
+  try:
```

Comment:
We can probably get rid of this try / except, since the one in handleStream does the same thing
(may be valid for other try / excepts)